test/bridge.go

```diff
@@ -0,0 +1,196 @@
+package test
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+)
+
+// bridgeConn is a net.Conn that represents an endpoint of the bridge.
+type bridgeConn struct {
+	br     *Bridge
+	id     int
+	readCh chan []byte
+}
+
+// Read reads data, block until the data becomes available.
+func (conn *bridgeConn) Read(b []byte) (int, error) {
+	if data, ok := <-conn.readCh; ok {
+		n := copy(b, data)
+		return n, nil
+	}
+	return 0, fmt.Errorf("bridgeConn closed")
+}
+
+// Write writes data to the bridge.
+func (conn *bridgeConn) Write(b []byte) (int, error) {
+	n := len(b)
+	conn.br.Push(b, conn.id)
+	return n, nil
+}
+
+// Close closes the bridge (releases resources used).
+func (conn *bridgeConn) Close() error {
+	close(conn.readCh)
+	return nil
+}
+
+// LocalAddr is not used
+func (conn *bridgeConn) LocalAddr() net.Addr { return nil }
+
+// RemoteAddr is not used
+func (conn *bridgeConn) RemoteAddr() net.Addr { return nil }
+
+// SetDeadline is not used
+func (conn *bridgeConn) SetDeadline(t time.Time) error { return nil }
+
+// SetReadDeadline is not used
+func (conn *bridgeConn) SetReadDeadline(t time.Time) error { return nil }
+
+// SetWriteDeadline is not used
+func (conn *bridgeConn) SetWriteDeadline(t time.Time) error { return nil }
+
+// Bridge represents a network between the two endpoints.
+type Bridge struct {
+	mutex sync.RWMutex
+	conn0 *bridgeConn
+	conn1 *bridgeConn
+
+	queue0to1 [][]byte
+	queue1to0 [][]byte
+}
+
+func inverse(s [][]byte) error {
+	if len(s) < 2 {
+		return fmt.Errorf("inverse requires more than one item in the array")
+	}
+
+	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
+		s[i], s[j] = s[j], s[i]
+	}
+	return nil
+}
+
+// drop n packets from the slice starting from offset
+func drop(s [][]byte, offset, n int) [][]byte {
+	if offset+n > len(s) {
+		n = len(s) - offset
+	}
+	return append(s[:offset], s[offset+n:]...)
+}
+
+// NewBridge creates a new bridge with two endpoints.
+func NewBridge() *Bridge {
+	br := &Bridge{
+		queue0to1: make([][]byte, 0),
+		queue1to0: make([][]byte, 0),
+	}
+
+	br.conn0 = &bridgeConn{
+		br:     br,
+		id:     0,
+		readCh: make(chan []byte),
+	}
+	br.conn1 = &bridgeConn{
+		br:     br,
+		id:     1,
+		readCh: make(chan []byte),
+	}
+
+	return br
+}
+
+// GetConn0 returns an endpoint of the bridge, conn0.
+func (br *Bridge) GetConn0() net.Conn {
+	return br.conn0
+}
+
+// GetConn1 returns an endpoint of the bridge, conn1.
+func (br *Bridge) GetConn1() net.Conn {
+	return br.conn1
+}
+
+// Push pushes a packet into the specified queue.
+func (br *Bridge) Push(d []byte, fromID int) {
+	br.mutex.Lock()
+	defer br.mutex.Unlock()
+
+	if fromID == 0 {
+		br.queue0to1 = append(br.queue0to1, d)
+	} else {
+		br.queue1to0 = append(br.queue1to0, d)
+	}
+}
+
+// Reorder inverses the order of packets currently in the specified queue.
+func (br *Bridge) Reorder(fromID int) error {
+	br.mutex.Lock()
+	defer br.mutex.Unlock()
+
+	var err error
+
+	if fromID == 0 {
+		err = inverse(br.queue0to1)
```

Comment:
can we just `return inverse(br.queue0to1)` here? I don't think we need to assign this to a var